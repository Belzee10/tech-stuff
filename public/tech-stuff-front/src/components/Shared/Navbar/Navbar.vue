<template>
  <fragment>
    <v-app-bar class="px-10" app color="primary" dark dense tile>
      <router-link to="/">
        <div class="mr-2 d-flex">
          <v-img
            to="home"
            src="../../../assets/logo.png"
            height="30"
            width="30"
            contain
          />
          <v-toolbar-title>{{ appTitle }}</v-toolbar-title>
        </div>
      </router-link>

      <v-spacer></v-spacer>
      <div class="hidden-md-and-down">
        <v-menu v-if="user" offset-y>
          <template v-slot:activator="{ on }">
            <v-btn text class="btn-user text-lowercase" v-on="on">
              {{ user.email }}
            </v-btn>
          </template>

          <v-list class="user-options">
            <v-list-item
              v-for="(item, index) in getUserOptions"
              :key="index"
              @click="() => goTo(item.link)"
            >
              <v-list-item-title>{{ item.title }}</v-list-item-title>
            </v-list-item>
            <v-divider></v-divider>
            <v-list-item class="logout" @click="handleLogout">
              <v-list-item-title>Logout</v-list-item-title>
            </v-list-item>
          </v-list>
        </v-menu>
        <div v-else>
          <v-btn text small :to="{ name: 'login' }">Sign in</v-btn>
          <span class="ml-1 mr-2">or</span>
          <v-btn :to="{ name: 'register' }" class="ma-2" outlined small
            >Sign up</v-btn
          >
        </div>
      </div>
      <v-app-bar-nav-icon
        class="hidden-md-and-up"
        @click="toggleDraw"
      ></v-app-bar-nav-icon>
    </v-app-bar>

    <v-navigation-drawer v-model="showDraw" class="primary" dark right app>
      <v-list-item>
        <v-list-item-content>
          <v-list-item-title class="title">
            {{ appTitle }}
          </v-list-item-title>
        </v-list-item-content>
      </v-list-item>

      <v-divider></v-divider>

      <v-list v-if="user" dense nav>
        <v-list-item v-for="(item, index) in getUserOptions" :key="index" link>
          <v-list-item-content>
            <v-list-item-title>{{ item.title }}</v-list-item-title>
          </v-list-item-content>
        </v-list-item>
        <v-divider></v-divider>
        <v-list-item class="logout" @click="handleLogout">
          <v-list-item-title>Logout</v-list-item-title>
        </v-list-item>
      </v-list>
      <v-list v-else dense nav>
        <v-list-item link :to="{ name: 'login' }">
          <v-list-item-content>
            <v-list-item-title>Sign in</v-list-item-title>
          </v-list-item-content>
        </v-list-item>
        <v-list-item link :to="{ name: 'register' }">
          <v-list-item-content>
            <v-list-item-title>Sign up</v-list-item-title>
          </v-list-item-content>
        </v-list-item>
      </v-list>
    </v-navigation-drawer>
  </fragment>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';

export default {
  name: 'Navbar',
  props: {
    appTitle: {
      type: String,
      default: ''
    }
  },
  data: () => ({
    showDraw: false,
    userOptions: [
      {
        title: 'My Orders',
        onlyAdmin: false,
        link: ''
      },
      {
        title: 'Manage Orders',
        onlyAdmin: true,
        link: ''
      },
      {
        title: 'Manage Categories',
        onlyAdmin: true,
        link: ''
      },
      {
        title: 'Manage Products',
        onlyAdmin: true,
        link: ''
      },
      {
        title: 'Manage Users',
        onlyAdmin: true,
        link: ''
      }
    ]
  }),
  computed: {
    ...mapGetters(['user']),
    getUserOptions() {
      return this.user.role === 'member'
        ? this.userOptions.filter(item => !item.onlyAdmin)
        : this.userOptions;
    }
  },
  methods: {
    ...mapActions(['logout']),
    toggleDraw() {
      this.showDraw = !this.showDraw;
    },
    goTo(link) {
      this.$router.push({ name: link });
    },
    handleLogout() {
      this.logout().then(() =>
        this.$router.push({ name: 'home' }).catch(() => {})
      );
    }
  }
};
</script>

<style lang="scss" scoped>
.v-toolbar__title {
  color: white;
}
</style>
