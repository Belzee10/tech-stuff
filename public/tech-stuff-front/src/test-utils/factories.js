import { mount, createLocalVue } from '@vue/test-utils';
import mergeWith from 'lodash.mergewith';
import Vuetify from 'vuetify';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);

let vuetify;

beforeEach(() => {
  vuetify = new Vuetify();
});

const customizer = (_, srcValue) => {
  if (Array.isArray(srcValue)) {
    return srcValue;
  }
  if (srcValue instanceof Object && Object.keys(srcValue).length === 0) {
    return srcValue;
  }
};

/**
 * create component wrapper
 * @param {*} Component
 * @param {Object} overrides
 */
const createWrapper = (Component, overrides) => {
  const defaultMountingOptions = {
    localVue,
    vuetify
  };
  return mount(
    Component,
    mergeWith(defaultMountingOptions, overrides, customizer)
  );
};

/**
 * create Vuex store
 * @param {Object} overrides
 */
const createStore = overrides => {
  const defaultStoreConfig = {};
  return new Vuex.Store(mergeWith(defaultStoreConfig, overrides, customizer));
};

export { createWrapper, createStore };
