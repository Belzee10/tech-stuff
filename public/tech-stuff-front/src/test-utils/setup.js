import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.config.productionTip = false;

Vue.use(Vuetify);

/**
 * silent "console.error"
 */
global.console = {
  error: jest.fn(),
  log: console.log,
  warn: console.warn,
  info: console.info,
  debug: console.debug
};
