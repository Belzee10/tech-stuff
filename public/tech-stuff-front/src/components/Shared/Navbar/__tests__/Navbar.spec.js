import flushPromises from 'flush-promises';
import { createWrapper, createStore } from '@/test-utils/factories';
import Navbar from '../Navbar.vue';

describe('Navbar.vue', () => {
  test('should render only the "member" options', () => {
    const user = {
      role: 'member'
    };
    const store = createStore({
      getters: {
        user: () => user
      }
    });
    const wrapper = createWrapper(Navbar, {
      store
    });
    wrapper.find('.btn-user').trigger('click');
    const options = wrapper.findAll('.user-options .v-list-item');
    expect(options).toHaveLength(2);
  });

  test('should render "all" the user options', () => {
    const user = {
      role: 'admin'
    };
    const store = createStore({
      getters: {
        user: () => user
      }
    });
    const wrapper = createWrapper(Navbar, {
      store
    });
    wrapper.find('.btn-user').trigger('click');
    const options = wrapper.findAll('.user-options .v-list-item');
    expect(options).toHaveLength(6);
  });

  test('should dispatch a "logout" action and redirect to home', async () => {
    const user = {
      role: 'member'
    };
    const mocks = {
      $router: {
        push: jest.fn(() => Promise.reject())
      }
    };
    const store = createStore({
      getters: {
        user: () => user
      }
    });
    store.dispatch = jest.fn(() => Promise.resolve());
    const wrapper = createWrapper(Navbar, {
      store,
      mocks
    });
    wrapper.find('.btn-user').trigger('click');
    const logout = wrapper.find('.user-options .logout');
    expect(logout.text()).toBe('Logout');
    logout.trigger('click');
    expect(store.dispatch).toHaveBeenCalledWith('logout');
    await flushPromises();
    expect(mocks.$router.push).toHaveBeenCalledWith({ name: 'home' });
  });

  test.todo('should render correctly');
});
