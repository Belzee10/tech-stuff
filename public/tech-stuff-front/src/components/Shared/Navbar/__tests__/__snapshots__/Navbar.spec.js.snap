// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar.vue should render correctly 1`] = `
<fragment>
  <header
    class="px-10 v-sheet v-sheet--tile theme--dark v-toolbar v-toolbar--dense v-app-bar v-app-bar--fixed primary"
    style="height: 48px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
  >
    <div
      class="v-toolbar__content"
      style="height: 48px;"
    >
      <router-link
        to="/"
      >
        <div
          class="mr-2 d-flex"
        >
          <div
            class="v-responsive v-image"
            style="height: 30px; width: 30px;"
            to="home"
          >
            <div
              class="v-image__image v-image__image--preload v-image__image--contain"
              mode="in-out"
              name="fade-transition"
              style="background-position: center center;"
            />
            <div
              class="v-responsive__content"
            />
          </div>
           
          <div
            class="v-toolbar__title"
          >
            
          </div>
        </div>
      </router-link>
       
      <div
        class="spacer"
      />
       
      <div
        class="hidden-md-and-down"
      >
        <div
          class="v-menu"
        >
          <button
            class="btn-user text-lowercase v-btn v-btn--flat v-btn--text theme--dark v-size--default"
            type="button"
          >
            <span
              class="v-btn__content"
            >
              
            
            
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-menu-down theme--dark"
                style="font-size: 16px;"
              />
            </span>
          </button>
          <div
            class="v-menu__content theme--light "
            role="menu"
            style="max-height: auto; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          />
        </div>
      </div>
       
      <button
        class="v-app-bar__nav-icon hidden-md-and-up v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-menu theme--dark"
          />
        </span>
      </button>
    </div>
  </header>
   
  <nav
    class="primary v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile v-navigation-drawer--right theme--dark"
    style="height: 100vh; top: 0px; transform: translateX(100%); width: 256px;"
  >
    <div
      class="v-navigation-drawer__content"
    >
      <div
        class="v-list-item theme--dark"
        tabindex="-1"
      >
        <div
          class="v-list-item__content"
        >
          <div
            class="v-list-item__title title"
          >
            
          
        
          </div>
        </div>
      </div>
       
      <hr
        aria-orientation="horizontal"
        class="v-divider theme--dark"
        role="separator"
      />
       
      <div
        class="v-list v-sheet v-sheet--tile theme--dark v-list--dense v-list--nav"
      >
        <div
          class="v-list-item v-list-item--link theme--dark"
          tabindex="0"
        >
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__title"
            >
              My Orders
            </div>
          </div>
        </div>
         
        <hr
          aria-orientation="horizontal"
          class="v-divider theme--dark"
          role="separator"
        />
         
        <div
          class="logout v-list-item v-list-item--link theme--dark"
          tabindex="0"
        >
          <div
            class="v-list-item__title"
          >
            Logout
          </div>
        </div>
      </div>
    </div>
    <div
      class="v-navigation-drawer__border"
    />
  </nav>
</fragment>
`;
